Security: All of miniupnpd runs with elevated privileges

`miniupnpd` must have elevated privileges (or at least `CAP_NET_ADMIN` and `CAP_NET_RAW` on Linux) to manipulate network rules, whatever OS it is running under.

It is common for daemons to start execution with superuser privileges, do whatever setup needs those privileges (e.g. opening/binding to ports < 1024), and then drop the elevated privileges before handling client interaction. In miniupnpd's case, it must retain some elevated privileges to be able to manipulate the system's networking rules dynamically.

From a security & robustness perspective, it is desirable to run code that handles external input (from the user or a network) with the minimum privileges possible. That way, should an attacker find an exploitable bug, they haven't gained any elevated privileges.

The common core of `miniupnpd` contains the majority of the complexity and presents the potential attack surface to the network. Thus it is highly desirable to run the core code with the minimal set of privileges required by the OS-specific network rule manipulation. Thus if the core is compromised, the attacker only gains a few elevated networking privileges, not the complete set root normally has.

In INSTALL.txt, is this still valid?

```
======================== Linux/netfilter nftables =========================

work is in progress. To build :
 > make -f Makefile.linux_nft
```

I am getting a missing dependency:

```
miniupnpd.c:49:10: fatal error: backtrace.h: No such file or directory
 #include <backtrace.h>
          ^~~~~~~~~~~~~
```

`libbacktrace` seems to be part of `GCC`?

```bash
$  git clone https://github.com/paul-chambers/miniupnp.git
Cloning into 'miniupnp'...
remote: Enumerating objects: 96, done.
remote: Counting objects: 100% (96/96), done.
remote: Compressing objects: 100% (57/57), done.
remote: Total 8852 (delta 62), reused 67 (delta 39), pack-reused 8756
Receiving objects: 100% (8852/8852), 3.16 MiB | 4.74 MiB/s, done.
Resolving deltas: 100% (6061/6061), done.
$ cd miniupnp/miniupnpd/
$ ./genconfig.sh
./genconfig.sh: line 364: /sbin/sysctl: No such file or directory
Configuring compilation for [Gentoo] [2.6] with [nftables] firewall software.
Please edit config.h for more compilation options.
$ sed -i 's/\/sbin\/sysctl/sysctl/g' genconfig.sh
$ ./genconfig.sh
Configuring compilation for [Gentoo] [2.6] with [nftables] firewall software.
Please edit config.h for more compilation options.
$ make -f Makefile.linux_nft
cc -O -g  -fno-strict-aliasing -fno-common -Wall -Wextra -Wstrict-prototypes -Wdeclaration-after-statement -DMINIUPNPD_GIT_REF=\"master-272e291\"  -D_GNU_SOURCE  -c -o miniupnpd.o miniupnpd.c
miniupnpd.c:49:10: fatal error: backtrace.h: No such file or directory
 #include <backtrace.h>
          ^~~~~~~~~~~~~
compilation terminated.
make: *** [<builtin>: miniupnpd.o] Error 1
$ sed -i 's/\#define HAS_BACKTRACE 1//g' config.h
$ make -f Makefile.linux_nft
...
$ make -f Makefile.linux_nft install
```

It seems like it's working for Transmission and Plex (below):

```bash
$ nft list ruleset|grep 32400
		iif "eno0" @th,16,16 32400 @nh,128,32 3232236053 @nh,72,8 6 accept
		iif "eno0" tcp dport 23647 dnat to 192.168.2.21:32400
```

Rergarding `backtrace.h`, on `Gentoo` its path appears to be `/usr/include/isc/backtrace.h`. So when I change the include to `#include <isc/backtrace.h>` in `miniupnpd.c` and `upnpglobalvars.h`, it does not complain about the import anymore. Compilation will still fail though:

```bash
make -f Makefile.linux_nft
cc -O -g  -fno-strict-aliasing -fno-common -Wall -Wextra -Wstrict-prototypes -Wdeclaration-after-statement -DMINIUPNPD_GIT_REF=\"master-bcba3d3\"  -D_GNU_SOURCE  -c -o miniupnpd.o miniupnpd.c
miniupnpd.c: In function '_log_error':
miniupnpd.c:167:3: warning: implicit declaration of function 'backtrace_full' [-Wimplicit-function-declaration]
   backtrace_full( backtrace_handle, 1, _cb_backtrace, NULL, NULL);
   ^~~~~~~~~~~~~~
miniupnpd.c: In function 'main':
miniupnpd.c:3029:21: warning: implicit declaration of function 'backtrace_create_state' [-Wimplicit-function-declaration]
  backtrace_handle = backtrace_create_state(argv[0], 0, NULL, NULL);
                     ^~~~~~~~~~~~~~~~~~~~~~
miniupnpd.c:3029:19: warning: assignment to 'struct backtrace_state *' from 'int' makes pointer from integer without a cast [-Wint-conversion]
  backtrace_handle = backtrace_create_state(argv[0], 0, NULL, NULL);
                   ^
cc -O -g  -fno-strict-aliasing -fno-common -Wall -Wextra -Wstrict-prototypes -Wdeclaration-after-statement -DMINIUPNPD_GIT_REF=\"master-bcba3d3\"  -D_GNU_SOURCE  -c -o upnphttp.o upnphttp.c
cc -O -g  -fno-strict-aliasing -fno-common -Wall -Wextra -Wstrict-prototypes -Wdeclaration-after-statement -DMINIUPNPD_GIT_REF=\"master-bcba3d3\"  -D_GNU_SOURCE  -c -o upnpdescgen.o upnpdescgen.c
cc -O -g  -fno-strict-aliasing -fno-common -Wall -Wextra -Wstrict-prototypes -Wdeclaration-after-statement -DMINIUPNPD_GIT_REF=\"master-bcba3d3\"  -D_GNU_SOURCE  -c -o upnpsoap.o upnpsoap.c
cc -O -g  -fno-strict-aliasing -fno-common -Wall -Wextra -Wstrict-prototypes -Wdeclaration-after-statement -DMINIUPNPD_GIT_REF=\"master-bcba3d3\"  -D_GNU_SOURCE  -c -o portinuse.o portinuse.c
cc -O -g  -fno-strict-aliasing -fno-common -Wall -Wextra -Wstrict-prototypes -Wdeclaration-after-statement -DMINIUPNPD_GIT_REF=\"master-bcba3d3\"  -D_GNU_SOURCE  -c -o upnpredirect.o upnpredirect.c
cc -O -g  -fno-strict-aliasing -fno-common -Wall -Wextra -Wstrict-prototypes -Wdeclaration-after-statement -DMINIUPNPD_GIT_REF=\"master-bcba3d3\"  -D_GNU_SOURCE  -c -o getifaddr.o getifaddr.c
cc -O -g  -fno-strict-aliasing -fno-common -Wall -Wextra -Wstrict-prototypes -Wdeclaration-after-statement -DMINIUPNPD_GIT_REF=\"master-bcba3d3\"  -D_GNU_SOURCE  -c -o daemonize.o daemonize.c
cc -O -g  -fno-strict-aliasing -fno-common -Wall -Wextra -Wstrict-prototypes -Wdeclaration-after-statement -DMINIUPNPD_GIT_REF=\"master-bcba3d3\"  -D_GNU_SOURCE  -c -o options.o options.c
cc -O -g  -fno-strict-aliasing -fno-common -Wall -Wextra -Wstrict-prototypes -Wdeclaration-after-statement -DMINIUPNPD_GIT_REF=\"master-bcba3d3\"  -D_GNU_SOURCE  -c -o upnppermissions.o upnppermissions.c
cc -O -g  -fno-strict-aliasing -fno-common -Wall -Wextra -Wstrict-prototypes -Wdeclaration-after-statement -DMINIUPNPD_GIT_REF=\"master-bcba3d3\"  -D_GNU_SOURCE  -c -o minissdp.o minissdp.c
cc -O -g  -fno-strict-aliasing -fno-common -Wall -Wextra -Wstrict-prototypes -Wdeclaration-after-statement -DMINIUPNPD_GIT_REF=\"master-bcba3d3\"  -D_GNU_SOURCE  -c -o natpmp.o natpmp.c
cc -O -g  -fno-strict-aliasing -fno-common -Wall -Wextra -Wstrict-prototypes -Wdeclaration-after-statement -DMINIUPNPD_GIT_REF=\"master-bcba3d3\"  -D_GNU_SOURCE  -c -o pcpserver.o pcpserver.c
cc -O -g  -fno-strict-aliasing -fno-common -Wall -Wextra -Wstrict-prototypes -Wdeclaration-after-statement -DMINIUPNPD_GIT_REF=\"master-bcba3d3\"  -D_GNU_SOURCE  -c -o upnpglobalvars.o upnpglobalvars.c
cc -O -g  -fno-strict-aliasing -fno-common -Wall -Wextra -Wstrict-prototypes -Wdeclaration-after-statement -DMINIUPNPD_GIT_REF=\"master-bcba3d3\"  -D_GNU_SOURCE  -c -o upnpevents.o upnpevents.c
cc -O -g  -fno-strict-aliasing -fno-common -Wall -Wextra -Wstrict-prototypes -Wdeclaration-after-statement -DMINIUPNPD_GIT_REF=\"master-bcba3d3\"  -D_GNU_SOURCE  -c -o upnputils.o upnputils.c
cc -O -g  -fno-strict-aliasing -fno-common -Wall -Wextra -Wstrict-prototypes -Wdeclaration-after-statement -DMINIUPNPD_GIT_REF=\"master-bcba3d3\"  -D_GNU_SOURCE  -c -o upnpstun.o upnpstun.c
cc -O -g  -fno-strict-aliasing -fno-common -Wall -Wextra -Wstrict-prototypes -Wdeclaration-after-statement -DMINIUPNPD_GIT_REF=\"master-bcba3d3\"  -D_GNU_SOURCE  -c -o upnppinhole.o upnppinhole.c
cc -O -g  -fno-strict-aliasing -fno-common -Wall -Wextra -Wstrict-prototypes -Wdeclaration-after-statement -DMINIUPNPD_GIT_REF=\"master-bcba3d3\"  -D_GNU_SOURCE  -c -o pcplearndscp.o pcplearndscp.c
cc -O -g  -fno-strict-aliasing -fno-common -Wall -Wextra -Wstrict-prototypes -Wdeclaration-after-statement -DMINIUPNPD_GIT_REF=\"master-bcba3d3\"  -D_GNU_SOURCE  -c -o linux/getifstats.o linux/getifstats.c
cc -O -g  -fno-strict-aliasing -fno-common -Wall -Wextra -Wstrict-prototypes -Wdeclaration-after-statement -DMINIUPNPD_GIT_REF=\"master-bcba3d3\"  -D_GNU_SOURCE  -c -o linux/ifacewatcher.o linux/ifacewatcher.c
cc -O -g  -fno-strict-aliasing -fno-common -Wall -Wextra -Wstrict-prototypes -Wdeclaration-after-statement -DMINIUPNPD_GIT_REF=\"master-bcba3d3\"  -D_GNU_SOURCE  -c -o netfilter_nft/nftnlrdr.o netfilter_nft/nftnlrdr.c
cc -O -g  -fno-strict-aliasing -fno-common -Wall -Wextra -Wstrict-prototypes -Wdeclaration-after-statement -DMINIUPNPD_GIT_REF=\"master-bcba3d3\"  -D_GNU_SOURCE  -c -o netfilter_nft/nftpinhole.o netfilter_nft/nftpinhole.c
cc -O -g  -fno-strict-aliasing -fno-common -Wall -Wextra -Wstrict-prototypes -Wdeclaration-after-statement -DMINIUPNPD_GIT_REF=\"master-bcba3d3\"  -D_GNU_SOURCE  -c -o netfilter_nft/nftnlrdr_misc.o netfilter_nft/nftnlrdr_misc.c
cc    miniupnpd.o upnphttp.o upnpdescgen.o upnpsoap.o upnpreplyparse.o minixml.o portinuse.o upnpredirect.o getifaddr.o daemonize.o options.o upnppermissions.o minissdp.o natpmp.o pcpserver.o upnpglobalvars.o upnpevents.o upnputils.o getconnstatus.o upnpstun.o upnppinhole.o pcplearndscp.o asyncsendto.o linux/getifstats.o linux/ifacewatcher.o linux/getroute.o netfilter_nft/nftnlrdr.o netfilter_nft/nftpinhole.o netfilter_nft/nfct_get.o netfilter_nft/nftnlrdr_misc.o  -lnftnl -lmnl -lssl -lcrypto -lz -ldl -luuid -lcap-ng -o miniupnpd
/usr/lib/gcc/x86_64-pc-linux-gnu/8.3.0/../../../../x86_64-pc-linux-gnu/bin/ld: miniupnpd.o: in function `_log_error':
/root/testing/miniupnp/miniupnpd/miniupnpd.c:167: undefined reference to `backtrace_full'
/usr/lib/gcc/x86_64-pc-linux-gnu/8.3.0/../../../../x86_64-pc-linux-gnu/bin/ld: miniupnpd.o: in function `main':
/root/testing/miniupnp/miniupnpd/miniupnpd.c:3029: undefined reference to `backtrace_create_state'
collect2: error: ld returned 1 exit status
make: *** [<builtin>: miniupnpd] Error 1
```

--- a/.travis.yml
+++ b/.travis.yml
@@ -14,6 +14,7 @@ addons:
     - libmnl-dev
     - libnftnl-dev
     - uuid-dev
+    - libcap-dev
     - cmake
     - doxygen
     - graphviz

--- a/miniupnpd/Changelog.txt
+++ b/miniupnpd/Changelog.txt
@@ -1,7 +1,14 @@
-$Id: Changelog.txt,v 1.463 2020/04/21 21:20:04 nanard Exp $
+$Id: Changelog.txt,v 1.464 2020/05/10 17:57:56 nanard Exp $
+
+2020/05/07:
+  Linux: use libcap or libcap-ng to drop unnecessary capabilities
 
 2020/05/04:
   Move configuration detection from Makefiles to configure (was genconfig.sh)
+  update linux Makefiles
+
+2020/04/29:
+  fix for bridges
 
 2020/04/21:
   Remove FW API detecting code from Makefile (BSD). generate bsdmake.inc

--- a/miniupnpd/Makefile.linux_nft
+++ b/miniupnpd/Makefile.linux_nft
@@ -1,4 +1,5 @@
 # MiniUPnP project
+# (c) 2018-2020 Thomas Bernard
 # (c) 2015 Tomofumi Hayashi
 # http://miniupnp.free.fr/ or https://miniupnp.tuxfamily.org/
 # Author : Tomofumi Hayashi

--- a/miniupnpd/Makefile.linux_nft
+++ b/miniupnpd/Makefile.linux_nft
@@ -18,14 +18,17 @@
 #
 #
 CONFIG_OPTIONS += --firewall=nftables
-CFLAGS = -O -g #-DDEBUG
+#CFLAGS = -O -g #-DDEBUG
 CFLAGS ?= -Os
 CFLAGS += -fno-strict-aliasing
 CFLAGS += -fno-common
+CFLAGS += -fstack-protector -fPIE
+CFLAGS += -D_FORTIFY_SOURCE=2
 CPPFLAGS += -D_GNU_SOURCE
 CFLAGS += -Wall
 CFLAGS += -Wextra -Wstrict-prototypes -Wdeclaration-after-statement
 #CFLAGS += -Wno-missing-field-initializers
+LDFLAGS += -Wl,-z,now -Wl,-z,relro -pie
 CC ?= gcc
 RM = rm -f
 INSTALL = install

--- a/miniupnpd/Makefile.macosx
+++ b/miniupnpd/Makefile.macosx
@@ -1,5 +1,5 @@
 # MiniUPnP project
-# http://miniupnp.free.fr/
+# http://miniupnp.free.fr/ or https://miniupnp.tuxfamily.org/
 # Author: Thomas Bernard
 # This Makefile should work for MacOSX
 #

--- a/miniupnpd/linux/getifstats.c
+++ b/miniupnpd/linux/getifstats.c
@@ -1,7 +1,7 @@
-/* $Id: getifstats.c,v 1.14 2018/03/13 23:05:21 nanard Exp $ */
+/* $Id: getifstats.c,v 1.16 2020/05/10 17:51:00 nanard Exp $ */
 /* MiniUPnP project
  * http://miniupnp.free.fr/ or http://miniupnp.tuxfamily.org/
- * (c) 2006-2019 Thomas Bernard
+ * (c) 2006-2020 Thomas Bernard
  * This software is subject to the conditions detailed
  * in the LICENCE file provided within the distribution */
 

--- a/miniupnpd/miniupnpd.c
+++ b/miniupnpd/miniupnpd.c
@@ -44,6 +44,9 @@
 /* for BSD's sysctl */
 #include <sys/sysctl.h>
 #endif
+#ifdef HAS_LIBCAP
+#include <sys/capability.h>
+#endif
 
 /* unix sockets */
 #ifdef USE_MINIUPNPDCTL
@@ -2325,6 +2328,49 @@ main(int argc, char * * argv)
 		return 1;
 	}
 #endif /* HAS_PLEDGE */
+#ifdef HAS_LIBCAP
+	{
+		cap_t caps = cap_get_proc();
+		if (caps == NULL) {
+			syslog(LOG_ERR, "cap_get_proc(): %m");
+		} else {
+			static const cap_value_t cap_list[3] = { CAP_NET_BROADCAST, CAP_NET_ADMIN, CAP_NET_RAW };
+			char * txt_caps = cap_to_text(caps, NULL);
+			if (txt_caps == NULL) {
+				syslog(LOG_ERR, "cap_to_text(): %m");
+			} else {
+				syslog(LOG_DEBUG, "capabilities %s", txt_caps);
+				if (cap_free(txt_caps) < 0) {
+					syslog(LOG_ERR, "cap_free(): %m");
+				}
+			}
+			if (cap_clear(caps) < 0) {
+				syslog(LOG_ERR, "cap_clear(): %m");
+			}
+			if (cap_set_flag(caps, CAP_PERMITTED, sizeof(cap_list)/sizeof(cap_list[0]), cap_list, CAP_SET) < 0) {
+				syslog(LOG_ERR, "cap_set_flag(): %m");
+			}
+			if (cap_set_flag(caps, CAP_EFFECTIVE, sizeof(cap_list)/sizeof(cap_list[0]), cap_list, CAP_SET) < 0) {
+				syslog(LOG_ERR, "cap_set_flag(): %m");
+			}
+			txt_caps = cap_to_text(caps, NULL);
+			if (txt_caps == NULL) {
+				syslog(LOG_ERR, "cap_to_text(): %m");
+			} else {
+				syslog(LOG_DEBUG, "capabilities %s", txt_caps);
+				if (cap_free(txt_caps) < 0) {
+					syslog(LOG_ERR, "cap_free(): %m");
+				}
+			}
+			if (cap_set_proc(caps) < 0) {
+				syslog(LOG_ERR, "cap_set_proc(): %m");
+			}
+			if (cap_free(caps) < 0) {
+				syslog(LOG_ERR, "cap_free(): %m");
+			}
+		}
+	}
+#endif /* HAS_LIBCAP */
 
 	/* main loop */
 	while(!quitting)

--- a/miniupnpd/miniupnpd.c
+++ b/miniupnpd/miniupnpd.c
@@ -47,6 +47,9 @@
 #ifdef HAS_LIBCAP
 #include <sys/capability.h>
 #endif
+#ifdef HAS_LIBCAP_NG
+#include <cap-ng.h>
+#endif
 
 /* unix sockets */
 #ifdef USE_MINIUPNPDCTL
@@ -2371,6 +2374,17 @@ main(int argc, char * * argv)
 		}
 	}
 #endif /* HAS_LIBCAP */
+#ifdef HAS_LIBCAP_NG
+	capng_setpid(getpid());
+	capng_clear(CAPNG_SELECT_BOTH);
+	if (capng_updatev(CAPNG_ADD, CAPNG_EFFECTIVE|CAPNG_PERMITTED, CAP_NET_BROADCAST, CAP_NET_ADMIN, CAP_NET_RAW, -1) < 0) {
+		syslog(LOG_ERR, "capng_updatev() failed");
+	} else {
+		if (capng_apply(CAPNG_SELECT_BOTH) < 0) {
+			syslog(LOG_ERR, "capng_apply() failed");
+		}
+	}
+#endif /* HAS_LIBCAP_NG */
 
 	/* main loop */
 	while(!quitting)

--- a/miniupnpd/netfilter/iptpinhole.c
+++ b/miniupnpd/netfilter/iptpinhole.c
@@ -1,7 +1,7 @@
-/* $Id: iptpinhole.c,v 1.20 2019/08/24 07:18:23 nanard Exp $ */
+/* $Id: iptpinhole.c,v 1.21 2020/05/10 17:49:33 nanard Exp $ */
 /* MiniUPnP project
  * http://miniupnp.free.fr/ or https://miniupnp.tuxfamily.org/
- * (c) 2012-2019 Thomas Bernard
+ * (c) 2012-2020 Thomas Bernard
  * This software is subject to the conditions detailed
  * in the LICENCE file provided within the distribution */
 

--- a/miniupnpd/netfilter_nft/nftnlrdr.c
+++ b/miniupnpd/netfilter_nft/nftnlrdr.c
@@ -1,8 +1,9 @@
 /*
  * vim: tabstop=4 shiftwidth=4 noexpandtab
  * MiniUPnP project
- * http://miniupnp.free.fr/ or http://miniupnp.tuxfamily.org/
+ * http://miniupnp.free.fr/ or https://miniupnp.tuxfamily.org/
  * (c) 2015 Tomofumi Hayashi
+ * (c) 2019 Sven Auhagen
  * (c) 2019 Paul Chambers
  *
  * This software is subject to the conditions detailed

--- a/miniupnpd/netfilter_nft/nftnlrdr_misc.c
+++ b/miniupnpd/netfilter_nft/nftnlrdr_misc.c
@@ -1,10 +1,10 @@
-/* $Id: nftnlrdr_misc.c,v 1.6 2019/10/02 22:00:10 nanard Exp $ */
+/* $Id: nftnlrdr_misc.c,v 1.8 2020/05/10 17:04:46 nanard Exp $ */
 /*
  * MiniUPnP project
  * http://miniupnp.free.fr/ or https://miniupnp.tuxfamily.org/
  * (c) 2015 Tomofumi Hayashi
- * (c) 2019 Thomas Bernard
  * (c) 2019 Paul Chambers
+ * (c) 2019-2020 Thomas Bernard
  *
  * This software is subject to the conditions detailed
  * in the LICENCE file provided within the distribution.

--- a/miniupnpd/netfilter_nft/nftpinhole.c
+++ b/miniupnpd/netfilter_nft/nftpinhole.c
@@ -1,7 +1,7 @@
-/* $Id: nftpinhole.c,v 1.3 2019/10/06 20:50:48 nanard Exp $ */
+/* $Id: nftpinhole.c,v 1.4 2020/05/10 17:01:49 nanard Exp $ */
 /* MiniUPnP project
  * http://miniupnp.free.fr/ or http://miniupnp.tuxfamily.org/
- * (c) 2012-2019 Thomas Bernard
+ * (c) 2012-2020 Thomas Bernard
  * This software is subject to the conditions detailed
  * in the LICENCE file provided within the distribution */
 

